File: codes/python/208.实现-trie-前缀树.py
class Trie:

	def __init__(self):
		self.childrens = [None] * 26
		self.is_end = False

	def insert(self, word: str) -> None:
		node = self
		for ch in word:
			index = ord(ch) - ord('a')
			if node.childrens[index] is None:
				node.childrens[index] = Trie()
			node = node.childrens[index]
		node.is_end = True

	def search(self, word: str) -> bool:
		node = self.searchPrefix(word)
		return node is not None and node.is_end

	def startsWith(self, prefix: str) -> bool:
		return self.searchPrefix(prefix) is not None

	def searchPrefix(self, prefix: str) -> 'Trie':
		node = self
		for ch in prefix:
			index = ord(ch) - ord('a')
			if node.childrens[index] is None:
				return None
			node = node.childrens[index]
		return node



# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: codes/java/208.实现-trie-前缀树.java
class Trie {

	private Trie[] childrens;
	private boolean isEnd;

	public Trie() {
		this.childrens = new Trie[26];
		this.isEnd = false;
	}

	public void insert(String word) {
		Trie iterator = this;
		char[] chs = word.toCharArray();
		for (int i = 0; i < chs.length; i++) {
			int index = chs[i] - 'a';
			if (iterator.childrens[index] == null) {
				iterator.childrens[index] = new Trie();
			}
			iterator = iterator.childrens[index];
		}
		iterator.isEnd = true;
	}

	public boolean search(String word) {
		Trie iterator = searchPrefix(word);
		return iterator != null && iterator.isEnd;
	}

	public boolean startsWith(String prefix) {
		return searchPrefix(prefix) != null;
	}

	private Trie searchPrefix(String prefix) {
		Trie iterator = this;
		char[] chs = prefix.toCharArray();
		for (int i = 0; i < chs.length; i++) {
			int index = chs[i] - 'a';
			if (iterator.childrens[index] == null) {
				return null;
			}
			iterator = iterator.childrens[index];
		}
		return iterator;
	}
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */